### ISSUE TITLE
ExpressionHelper contains a compatibility bug with Roslyn compiler

### ISSUE THREAD

**Issue by doggy8088** [aspnet/AspNetWebStack#117]
There is a bug in the ASP.NET MVC 5.2.3 for a long time.   It's hard to explain this in English.  So I made a few steps to reproduce this problem.

# STEPS TO REPRODUCE THE PROBLEM

1. Create a simple ASP.NET MVC 5 project using Visual Studio 2017
2. Apply change to the follow code snippets.  Only 3 files to change.
   https://gist.github.com/doggy8088/58dd3c773ef9877155b19ebb2fda8525
3. Start debugging and navigate to /Home/Index page.  Then you will see the result like below:
   ![image](https://user-images.githubusercontent.com/88981/34677022-c5b3ea6c-f4c9-11e7-8d81-fc553696ff4b.png)
As you can see, the `@Html.TextBoxFor(m => data[i].Name)` generated field name in the for-loop in the view are all wrong.  

# MORE INFO

This problem is only exists when ASP.NET MVC 5 project using Roslyn to compiler views.

I was wrote an article about 2 years ago.  My article is written in Chinse.  If some of you can read Chinese, you can also reference my article over here: https://blog.miniasp.com/post/2016/03/06/ASPNET-MVC-5-View-Roslyn-problem-workaround.aspx

I actually proposed a [pull request](https://aspnetwebstack.codeplex.com/SourceControl/network/forks/doggy8088/ExpressionHelperFixed/changeset/1f069a92741a58f6410fe1b0773154f5c37c9257) to fix this problem about 2 years ago too.  You may check it out too.

**Comment by mkArtakMSFT** [comment]
@dougbu, can you please look into this?

**Comment by IanKemp** [comment]
tl;dr backporting the fixes from [this commit](MVChttps://github.com/aspnet/Mvc/commit/a045324d3a7bd29f030a120d2edd90fedb3fb5be) in ASP.NET Core should resolve the issue.

**Comment by doggy8088** [comment]
@mkArtakMSFT Any new progress on this issue?

**Comment by mkArtakMSFT** [comment]
@dougbu, did you get a chance to look into this?

**Comment by dougbu** [comment]
@mkArtakMSFT I read the comments over the weekend and no longer think we need an investigation.

We fixed a related problem in ASP.NET Core and can do something similar here. That should fix the problem. Should however spend some time adding a number of tests covering more-complicated lambda expressions in this repo.

BTW, I'm surprised Roslyn versions are material here: I would expect the set of captured local variables to remain fairly consistent.

**Comment by dougbu** [comment]
b32ab42712
